### ISSUE TITLE
queue cache mechanism

### ISSUE THREAD

**Issue by creesch** [toolbox-team/reddit-moderator-toolbox#187]
Needed for our own overlays to figure out if things are filtered and for RR as it currently removes everything as from redesign it is not clear if things are removed or filtered when clicking "add removal reason".

Cache freshness will be determined by a `lastRefresh` timestamp which will be checked against the thing post or edited timestamp we are looking for.

**Comment by creesch** [comment]
Something along these lines (untested as I am behind a pc where I can't do dev stuff atm)

```javascript

// Obviously this is going to be a proper cache thing. 
let cacheObject = {
    LastRefresh: 1571815655569,
    subreddits: {
        toolbox: {
            lastRefresh: 1571815655569,
            things: ['t3_etc']
        }
    }
}


TBCore.queueCheck = function queueCheck(thingName, thingTimestamp, subreddit, allsubs) {
    return new Promise(async resolve => {
        // Check if we need to fetch data. 
        let lastRefresh = 0; 
        if (allsubs) {
            lastRefresh = cacheObject.LastRefresh;
        } else if (Object.prototype.hasOwnProperty.call(cacheObject.subreddits, subreddit)) {
            lastRefresh = cacheObject.subreddits[subreddit].lastRefresh;
        }

        // The thing timestamp is bigger than the last refresh or cache isn't fresh anymore.
        if (thingTimestamp * 1000 > lastRefresh || Date.now() - lastRefresh > 1000 * 60) {
            const updatedQueue = await TBApi.getJSON(`/r/${allsubs ? 'mod' : subreddit}/about/modqueue`);
            const nowRefresh = Date.now();
            if (allsubs) {
                cacheObject.LastRefresh = nowRefresh;
                cacheObject.subreddits = {};
            } else {
                cacheObject.subreddits[subreddit] = {
                    lastRefresh: nowRefresh,
                    things: []
                }
            }

            updatedQueue.data.children.forEach(thing => {
                if (Object.prototype.hasOwnProperty.call(cacheObject.subreddits, thing.data.subreddit)) {
                    cacheObject.subreddits[subreddit].things.push(thing.data.name)
                } else {
                    cacheObject.subreddits[thing.data.subreddit] = {
                        lastRefresh: nowRefresh,
                        things: [thing.data.name]
                    }
                }
            });
        }
        // Resolve to either true or false.
        resolve(cacheObject.subreddits[subreddit].things.includes(thingName));
    });
}
```

**Comment by k3n** [comment]
Commenting from the hip here, pardon if I'm missing something. I briefly looked for what `proper cache thing` refers to but couldn't find anything, but I'd be more inclined to moving more of the logic that interacts with the cache object into the cache object itself.

e.g.

```js
const lastRefresh = cache.getLastRefresh({ allSubs });
const lastRefresh = cache.getLastRefresh({ subreddit });
// ...
cache.updateLastRefresh({ allSubs });
cache.updateLastRefresh({ subreddit });
// ...
cache.upsertSubreddit(subreddit);
// ...
cache.checkSubreddit(subreddit);
```

The idea would be to minimize the code dealing with cache internals within `queueCheck()`.

**Comment by creesch** [comment]
I get what you are saying but that isn't really how caching (currently) works in toolbox. 

Due to legacy reasons we have a few variety of cache mechanism but the most simple one (and the one I was refering to is the storage one found here: https://github.com/toolbox-team/reddit-moderator-toolbox/blob/master/extension/data/tbstorage.js#L440-L484

Which basically stores it in the background page localstorage has it doesn't need to be reliably stored and as it is caching we want not stored next to the settings. 


Around that we have caching surrounding a few specific things where it expires on a timer, like here: https://github.com/toolbox-team/reddit-moderator-toolbox/blob/master/extension/data/tbstorage.js#L440-L484

https://github.com/toolbox-team/reddit-moderator-toolbox/blob/master/extension/data/background/handlers/cache.js

Which doesn't really need to apply here as we can and want to more dynamically determine freshness.

**Comment by creesch** [comment]
I have been thinking about this for a while now and really want to implement this at some point. Because currently flatview and the toolbox profile simply don't show if something is filtered and that makes them less useful.  The problem is that building those is a mostly synchronous event where checking the cache and possibly the queue is an async operation. 

So what I have been thinking is that what I am going to do is the following 

- Build the cache mechanism (checking and updating) as an async operation. 
- Always do it on a per subreddit basis. 
- Do this with a flag for "update in progress" queuing cache requests until the cache is updated from the modlog. 
- Make it so that the comment and post build utilities build and render things like they are now, then do a cache request and then once that resolves update the thing to add the three big buttons. 



tl;dr make checking async, build elements as per usual and update them as needed.
